File: services/be/app/utils/formatters.py
from __future__ import annotations

from datetime import datetime, timezone


def _indian_comma_grouping(n: int) -> str:
    """Format an integer with Indian comma grouping: 1,23,456."""
    s = str(n)
    if len(s) <= 3:
        return s
    # Last 3 digits, then groups of 2
    last3 = s[-3:]
    rest = s[:-3]
    groups = []
    while rest:
        groups.append(rest[-2:])
        rest = rest[:-2]
    groups.reverse()
    return ",".join(groups) + "," + last3


def format_inr(value: float, show_sign: bool = False) -> str:
    """Format as Indian Rupees: '₹1,23,456'."""
    abs_val = abs(value)
    sign = "+" if value >= 0 and show_sign else ("-" if value < 0 else "")
    formatted = _indian_comma_grouping(round(abs_val))
    return f"{sign}₹{formatted}"


def format_lakhs_crores(value: float) -> str:
    """Format in lakhs/crores: '₹12.34L' or '₹1.23Cr'."""
    abs_val = abs(value)
    sign = "-" if value < 0 else ""
    if abs_val >= 10_000_000:
        return f"{sign}₹{abs_val / 10_000_000:.2f}Cr"
    elif abs_val >= 100_000:
        return f"{sign}₹{abs_val / 100_000:.2f}L"
    elif abs_val >= 1000:
        return f"{sign}₹{abs_val / 1000:.1f}K"
    return f"{sign}₹{round(abs_val)}"


def format_percent(value: float, show_sign: bool = False) -> str:
    """Format percentage: '+5.25%'."""
    sign = "+" if value >= 0 and show_sign else ""
    return f"{sign}{value:.2f}%"


def time_ago(dt: datetime) -> str:
    """Human-readable relative time: '5m ago', '2h ago'."""
    now = datetime.now(timezone.utc)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    diff = now - dt
    diff_mins = int(diff.total_seconds() / 60)
    diff_hours = int(diff.total_seconds() / 3600)

    if diff_mins < 1:
        return "Just now"
    if diff_mins < 60:
        return f"{diff_mins}m ago"
    if diff_hours < 24:
        return f"{diff_hours}h ago"
    return dt.strftime("%d-%b")

File: services/fe/src/utils/formatters.js
/**
 * Format a number as Indian Rupees
 */
export function formatINR(value, showSign = false) {
  const absValue = Math.abs(value)
  const sign = value >= 0 ? (showSign ? '+' : '') : '-'

  const formatted = absValue.toLocaleString('en-IN', {
    maximumFractionDigits: 0,
  })

  return `${sign}₹${formatted}`
}

/**
 * Format a number in lakhs/crores
 */
export function formatLakhsCrores(value) {
  const absValue = Math.abs(value)
  const sign = value >= 0 ? '' : '-'

  if (absValue >= 10000000) {
    return `${sign}₹${(absValue / 10000000).toFixed(2)}Cr`
  } else if (absValue >= 100000) {
    return `${sign}₹${(absValue / 100000).toFixed(2)}L`
  } else if (absValue >= 1000) {
    return `${sign}₹${(absValue / 1000).toFixed(1)}K`
  }
  return `${sign}₹${absValue}`
}

/**
 * Format percentage
 */
export function formatPercent(value, showSign = false) {
  const sign = value >= 0 ? (showSign ? '+' : '') : ''
  return `${sign}${value.toFixed(2)}%`
}

/**
 * Get color class based on value (positive/negative)
 */
export function getPnlColor(value) {
  if (value > 0) return 'text-positive'
  if (value < 0) return 'text-negative'
  return 'text-slate-400'
}

/**
 * Get background color class based on value
 */
export function getPnlBgColor(value) {
  if (value > 0) return 'bg-positive/10'
  if (value < 0) return 'bg-negative/10'
  return 'bg-slate-500/10'
}

/**
 * Format time for display
 */
export function formatTime(date) {
  const d = new Date(date)
  return d.toLocaleTimeString('en-IN', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  })
}

/**
 * Format date for display
 */
export function formatDate(date) {
  const d = new Date(date)
  return d.toLocaleDateString('en-IN', {
    day: '2-digit',
    month: 'short',
  })
}

/**
 * Format timestamp for chat messages
 */
export function formatTimestamp(date) {
  const d = new Date(date)
  return d.toLocaleTimeString('en-IN', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
  })
}

/**
 * Format number with Indian comma separators
 */
export function formatNumber(value) {
  return value.toLocaleString('en-IN')
}

/**
 * Get severity color for alerts
 */
export function getSeverityColors(severity) {
  switch (severity) {
    case 'critical':
      return { text: 'text-negative', bg: 'bg-negative/20', border: 'border-negative/50' }
    case 'high':
      return { text: 'text-negative', bg: 'bg-negative/10', border: 'border-negative/30' }
    case 'medium':
      return { text: 'text-warning', bg: 'bg-warning/10', border: 'border-warning/30' }
    case 'low':
      return { text: 'text-accent', bg: 'bg-accent/10', border: 'border-accent/30' }
    default:
      return { text: 'text-slate-400', bg: 'bg-slate-500/10', border: 'border-slate-500/30' }
  }
}

/**
 * Calculate time ago string
 */
export function timeAgo(date) {
  const now = new Date()
  const d = new Date(date)
  const diffMs = now - d
  const diffMins = Math.floor(diffMs / 60000)
  const diffHours = Math.floor(diffMs / 3600000)

  if (diffMins < 1) return 'Just now'
  if (diffMins < 60) return `${diffMins}m ago`
  if (diffHours < 24) return `${diffHours}h ago`
  return formatDate(date)
}
